File: tools/erase-firmware.py
#!/usr/bin/env python3
# Tadpole — remove the installed system files, returning to a first-run state.
#
#   tools/erase-firmware.py [--really-delete]
#
# This is tools/erase-firmware.sh as one program, for the same reason
# install-firmware.py exists: Windows has no shell to run the .sh. The shell
# version remains the Linux entry point and is unchanged.
#
# For testing the setup wizard, which is otherwise awkward to reach once an
# install works — and, on Windows especially, for backing out of a half-done
# install without deleting anything you cannot get back.
#
# IT MOVES RATHER THAN DELETES, by default.
#
# Re-installing needs the firmware packages AND ubi_reader, and if either is
# missing a genuine delete leaves you unable to run anything until you sort
# that out. A rename costs nothing and is reversible, so the destructive
# version is opt-in.
#
# Never touches: games/ (your cartridge backups), sources/ (your firmware
# downloads), or anything you built.

import errno
import os
import shutil
import sys
import time

HERE = os.path.dirname(os.path.abspath(__file__))
PROJ = os.environ.get("TADPOLE_PROJECT") or os.path.dirname(HERE)


def rel(path):
    try:
        return os.path.relpath(path, PROJ)
    except ValueError:                      # different drive on Windows
        return path


def main(argv):
    hard = "--really-delete" in argv[1:]
    backup = os.path.join(PROJ, ".erased-" + time.strftime("%Y%m%d-%H%M%S"))
    moved = 0

    def takeaway(path):
        nonlocal moved
        if not os.path.exists(path):
            return
        if hard:
            shutil.rmtree(path, ignore_errors=True)
            print("  deleted  %s" % rel(path))
        else:
            # PRESERVE THE RELATIVE PATH inside the backup, so restoring is a
            # single copy. Flattening everything into one directory loses where
            # each piece came from — rootfs/<ver> and runtime/sysroot land side
            # by side and the restore command becomes a guess.
            dest = os.path.join(backup, os.path.dirname(rel(path)))
            os.makedirs(dest, exist_ok=True)
            try:
                shutil.move(path, os.path.join(dest, os.path.basename(path)))
            except OSError as e:
                # A file still open by a running guest cannot be renamed on
                # Windows, and it is the common way this fails. Say which one.
                print("  cannot move %s: %s" % (rel(path), e.strerror or e),
                      file=sys.stderr)
                if e.errno in (errno.EACCES, errno.EPERM):
                    print("  close Tadpole and any running title, then retry.",
                          file=sys.stderr)
                return
            print("  moved    %s" % rel(path))
        moved += 1

    print("Erasing installed system files...")

    # The extracted firmware.
    rootfs = os.path.join(PROJ, "rootfs")
    if os.path.isdir(rootfs):
        for name in sorted(os.listdir(rootfs)):
            d = os.path.join(rootfs, name)
            if os.path.isdir(d):
                takeaway(d)

    # The sysroot is GENERATED — links into rootfs plus the writable
    # directories the guest needs. Removing it is safe and it is rebuilt.
    takeaway(os.path.join(PROJ, "runtime", "sysroot"))

    if moved == 0:
        print("  nothing installed — already at a first-run state")
        return 0

    print()
    if hard:
        print("Deleted. Re-install with:  tools/install-firmware.py <LFC_Downloads>")
    else:
        print("Moved to: %s" % rel(backup))
        print("Restore by copying its contents back over the project directory.")
        print("Or re-install:  tools/install-firmware.py <LFC_Downloads>")
    print("Tadpole will show the setup wizard on next launch.")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: android/app/java/org/tadpole/view/EraseFirmware.java
package org.tadpole.view;

import java.io.File;
import java.io.PrintStream;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Date;
import java.util.Locale;

/**
 * tools/erase-firmware.py, in Java.
 *
 * <p>Removes the installed system files, returning to a first-run state, for
 * testing the setup wizard — which is otherwise awkward to reach once an
 * install works — and for backing out of a half-done install.
 *
 * <p>IT MOVES RATHER THAN DELETES, by default, and that is the whole design of
 * the original: re-installing needs the firmware packages and a UBIFS reader,
 * and on Android it needs a desktop and push-firmware.sh, so a genuine delete
 * can leave you with nothing to run and no quick way back. A rename costs
 * nothing and is reversible. --really-delete is the opt-in.
 *
 * <p>Never touches games/ (cartridge backups), sources/ (firmware downloads),
 * or anything built.
 *
 * <p>The one deliberate difference from the Python: it does not print the
 * "re-install with tools/install-firmware.py" line, because that tool is not
 * one this device has. It names push-firmware.sh instead, which is the route
 * that exists here.
 */
final class EraseFirmware implements Tools.Tool {

    @Override
    public boolean run(String[] argv, PrintStream out) throws Exception {
        boolean hard = Arrays.asList(argv).contains("--really-delete");
        String stamp = new SimpleDateFormat("yyyyMMdd-HHmmss", Locale.ROOT)
                .format(new Date());
        File backup = new File(Tools.proj(), ".erased-" + stamp);
        int[] moved = { 0 };

        out.println("Erasing installed system files...");

        /* The extracted firmware. Sorted, so the log reads the same twice. */
        File rootfs = Tools.rootfs();
        if (rootfs.isDirectory()) {
            File[] kids = rootfs.listFiles();
            if (kids != null) {
                Arrays.sort(kids);
                for (File d : kids)
                    if (d.isDirectory() && !takeaway(d, hard, backup, moved, out))
                        return false;
            }
        }

        /* The sysroot is GENERATED — links into rootfs plus the writable
         * directories the guest needs. Removing it is safe; it is rebuilt. */
        if (!takeaway(Tools.sysroot(), hard, backup, moved, out))
            return false;

        if (moved[0] == 0) {
            out.println("  nothing installed — already at a first-run state");
            return true;
        }

        out.println("");
        if (hard) {
            out.println("Deleted.");
        } else {
            out.println("Moved to: " + Tools.rel(backup));
            out.println("Restore by copying its contents back over " + Tools.rel(Tools.proj()) + ".");
        }
        out.println("Re-install from a desktop with android/push-firmware.sh.");
        out.println("Tadpole will show the setup wizard on next launch.");
        return true;
    }

    /** Returns false only on a failure worth stopping for. */
    private boolean takeaway(File path, boolean hard, File backup, int[] moved,
                             PrintStream out) {
        if (!path.exists()) return true;

        if (hard) {
            Tools.rmTree(path);
            out.println("  deleted  " + Tools.rel(path));
            moved[0]++;
            return true;
        }

        /* PRESERVE THE RELATIVE PATH inside the backup, so restoring is a
         * single copy. Flattening everything into one directory loses where
         * each piece came from — rootfs/<ver> and runtime/sysroot land side by
         * side and the restore becomes a guess. */
        String relPath = Tools.rel(path);
        int slash = relPath.lastIndexOf('/');
        File dest = slash < 0 ? backup : new File(backup, relPath.substring(0, slash));
        if (!dest.isDirectory() && !dest.mkdirs()) {
            out.println("  cannot create " + Tools.rel(dest));
            return false;
        }
        try {
            if (!Tools.move(path, new File(dest, path.getName()))) {
                out.println("  cannot move " + relPath);
                out.println("  stop the guest and any running title, then retry.");
                return false;
            }
        } catch (java.io.IOException e) {
            /* A file still held open by a RUNNING GUEST is the common way this
             * fails, and on Android the guest is a root process the app cannot
             * even signal — so say which one and what to do about it, rather
             * than letting the panel show a bare exception. */
            out.println("  cannot move " + relPath + ": " + e.getMessage());
            out.println("  stop the guest and any running title, then retry.");
            return false;
        }
        out.println("  moved    " + relPath);
        moved[0]++;
        return true;
    }
}
